articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,230 @@
+
+### Profiling Tools
+
+In addition to gathering performance statistics, it is useful to introspect the conditions under which the statistics were gathered to pinpoint areas where the code needs improvement.
+
+Two well-established open source command line profiling tools with orthogonal methodologies are:
+
+- `perf`: counts system-wide hardware events, e.g. instructions executed, cache misses, etc.
+
+  - Linux only (5).
+
+- `gprof`: recompiles source code with instrumentation to count time spent in each function for a particular program, with option to format output in a call graph.
+
+  - GNU and C/C++ only (6).
+
+However, these particular tools may not be well suited to an RTOS or a bare metal embedded system: `perf` is specialized for the Linux kernel and `gprof` requires code to be compiled and linked with `gcc`.
+These tools could be used to profile ROS 2 code on a patched Linux kernel like RT_PREEMPT, but there may be performance quirks in another RTOS that are not captured on Linux.
+When this proposal is implemented, if changes are made to improve performance is made based on profiling data from Linux but no overall improvement is seen in the embedded system, then an alternative approach to profiling should be investigated.
+For example, the RTEMS real-time operating system offers a tracing tool that could be used for profiling (7).
+
+### Verification
+
+In addition to performance testing, which can be thought of as a kind of regression testing, any change made to ROS 2 with the intent to improve performance in the real-time test system should be verified with an automated test.
+For example, the allocator pipeline has a test case for publishers/subscribers with custom allocators that fails if calls are made to the default system allocator during the real-time execution phase.
+A similar test could be written for the proposed goal of abstracting synchronization primitives, though the implementation may be more challenging since the allocator pipeline was helped by the ability to override global new in C++.
+
+The expected response of a hard or firm real-time system to missed deadlines should also be thoroughly tested.
+
+## Timeline
+
+This proposal represents a short burst of work to develop the initial test system, followed by extended maintenance as testing identifies performance bottlenecks.
+
+Since a key part of this proposal is the development of a hard real-time embedded system, this work depends on the development of `freertps` and `rmw_freertps` and the porting of `rclcpp` to STM32 or another microcontroller type.
+
+- Alpha 4 (2/12/15): Create initial system and test suite (2-3 weeks).
+- Every release thereafter: Iterate on ARPI, refactor and expand test suite as new features are added to core ROS 2 libraries (1-2 weeks each release cycle).
+
+This work should share as much labor as possible with the roscar effort (running ROS 2 on a small mobile robot) or perhaps be considered a subset of roscar.
```

Comment:
I wouldn't refer to `roscar` in this article. Also the point should be obvious - we always try to reuse stuff to minimize our effort.